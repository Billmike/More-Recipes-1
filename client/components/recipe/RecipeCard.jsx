import React from 'react';

import RecipeImage from '../common/RecipeImage';

const RecipeCard = () => (
  <div className="card">
    <div className="card-image">
      <RecipeImage />
    </div>
    <div className="card-action">
      <a id="favorite" className="chip waves-effect waves-red ">
        <i className="fa fa-heart" />
      </a>
      <a id="upvote" className="chip waves-effect waves-red">
        <i className="fa fa-thumbs-up" />
      </a>
      <a id="downvote" className="chip waves-effect waves-red">
        <i className="fa fa-thumbs-down" />
      </a>
    </div>
  </div>
);

export default RecipeCard;
