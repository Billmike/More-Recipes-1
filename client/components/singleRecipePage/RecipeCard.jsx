import React from 'react';
import PropTypes from 'prop-types';
import ReactTooltip from 'react-tooltip';

import RecipeImage from '../common/RecipeImage';
import { decodeToken } from '../../utils/authenticate';

const propTypes = {
  recipe: PropTypes.shape({
    title: PropTypes.string,
    recipeImage: PropTypes.string,
    upvotes: PropTypes.number,
    downvotes: PropTypes.number,
    favorites: PropTypes.number,
    views: PropTypes.views,
  }).isRequired,
  favoritedUsers: PropTypes.arrayOf(PropTypes.shape()).isRequired,
  voters: PropTypes.arrayOf(PropTypes.shape()).isRequired,
  voteMessage: PropTypes.string,
  favoriteMessage: PropTypes.string,
  handleUpvote: PropTypes.func.isRequired,
  handleDownvote: PropTypes.func.isRequired,
  handleFavorite: PropTypes.func.isRequired,
};

const defaultProps = {
  voteMessage: null,
  favoriteMessage: null,
};

/**
 * RecipeCard component
 *
 * @param {object} props - The properties passed to the component
 *
 * @returns {object} - React element
 */
const RecipeCard = (props) => {
  const { recipe, voters, favoritedUsers, voteMessage, favoriteMessage,
    handleUpvote, handleDownvote, handleFavorite } = props;
  const { title, recipeImage, upvotes, downvotes, favorites, views } = recipe;

  let userId;

  if (decodeToken() !== null) {
    const { user: { id } } = decodeToken();
    userId = id;
  }

  let isUserFavorite = favoritedUsers && favoritedUsers
    .some(user => user.userId === userId);
  let userUpvoted = voters && voters
    .some(user => user.userId === userId && user.hasVoted);
  let userDownvoted = voters && voters
    .some(user => user.userId === userId && !user.hasVoted);

  if (voteMessage && (voteMessage === 'Upvote recorded'
    || voteMessage === 'Upvote recorded and downvote removed')) {
    userUpvoted = true;
    userDownvoted = false;
  }
  if (voteMessage && (voteMessage === 'Downvote recorded'
    || voteMessage === 'Downvote recorded and upvote removed')) {
    userDownvoted = true;
    userUpvoted = false;
  }
  if (voteMessage && voteMessage === 'Upvote removed') {
    userUpvoted = false;
    userDownvoted = false;
  }
  if (voteMessage && voteMessage === 'Downvote removed') {
    userDownvoted = false;
    userUpvoted = false;
  }

  if (favoriteMessage && favoriteMessage === 'Recipe added to favorites') {
    isUserFavorite = true;
  }
  if (favoriteMessage && favoriteMessage === 'Recipe removed from favorites') {
    isUserFavorite = false;
  }

  return (
    <div className="card single-card">
      <div className="card-image">
        <RecipeImage title={title} recipeImage={recipeImage} />
      </div>
      <div className="card-action center-align">
        <a id="views">
          <i className="material-icons black-text" data-tip="views">
            visibility
          </i>
          <span className="black-text">{views}</span>
        </a>
        <a
          role="button"
          tabIndex="0"
          id="favorite"
          className={`chip waves-effect waves-red ${
            isUserFavorite ? 'red' : ''}`}
          onClick={handleFavorite}
          data-tip="favorite"
        >
          <i className={`fa fa-heart ${
            isUserFavorite ? 'white-text' : 'black-text'}`}
          />
          <span className={`${isUserFavorite ? 'white-text' : 'black-text'}`}>
            {favorites}
          </span>
        </a>
        <a
          role="button"
          tabIndex="0"
          id="upvote"
          className={`chip waves-effect waves-red ${userUpvoted ? 'red' : ''}`}
          onClick={handleUpvote}
          data-tip="upvote"
        >
          <i className={`fa fa-thumbs-up ${
            userUpvoted ? 'white-text' : 'black-text'}`}
          />
          <span className={`${userUpvoted ? 'white-text' : 'black-text'}`}>
            {upvotes}
          </span>
        </a>
        <a
          role="button"
          tabIndex="0"
          id="downvote"
          className={`chip waves-effect waves-red ${
            userDownvoted ? 'red' : ''}`}
          onClick={handleDownvote}
          data-tip="downvote"
        >
          <i className={`fa fa-thumbs-down ${
            userDownvoted ? 'white-text' : 'black-text'}`}
          />
          <span className={`${userDownvoted ? 'white-text' : 'black-text'}`}>
            {downvotes}
          </span>
        </a>
        <ReactTooltip />
      </div>
    </div>
  );
};

RecipeCard.propTypes = propTypes;
RecipeCard.defaultProps = defaultProps;

export default RecipeCard;
