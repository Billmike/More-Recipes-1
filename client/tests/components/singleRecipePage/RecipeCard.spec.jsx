import React from 'react';

import RecipeCard from '../../../components/singleRecipePage/RecipeCard';

describe('RecipeCard component', () => {
  it('should render component', () => {
    const props = {
      singleRecipe: {
        recipe: {
          title: 'Macaroni',
          recipeImage: 'Recipe image URL',
          upvotes: 4,
          downvotes: 0,
          favorites: 8,
          views: 25,
        },
        favoritedUsers: [],
        voters: [],
      },
      handleUpvote: jest.fn(),
      handleDownvote: jest.fn(),
      handleFavorite: jest.fn(),
    };
    const wrapper = shallow(<RecipeCard {...props} />);

    expect(wrapper).toBeDefined();
    expect(wrapper).toMatchSnapshot();
  });
  it('should render component with upvote', () => {
    const props = {
      singleRecipe: {
        recipe: {
          title: 'Macaroni',
          recipeImage: 'Recipe image URL',
          upvotes: 4,
          downvotes: 0,
          favorites: 8,
          views: 25,
        },
        favoritedUsers: [{
          userId: 1,
        }, {
          userId: 2,
        }],
        voters: [{
          userId: 1,
          hasVoted: true,
        }],
        voteMessage: 'Upvote recorded',
      },
      handleUpvote: jest.fn(),
      handleDownvote: jest.fn(),
      handleFavorite: jest.fn(),
    };
    const wrapper = shallow(<RecipeCard {...props} />);

    expect(wrapper).toBeDefined();
    wrapper.find('#upvote').simulate('click');
    expect(props.handleUpvote).toBeCalled();
    expect(wrapper).toMatchSnapshot();
  });
  it('should render component with upvote', () => {
    const props = {
      singleRecipe: {
        recipe: {
          title: 'Macaroni',
          recipeImage: 'Recipe image URL',
          upvotes: 4,
          downvotes: 0,
          favorites: 8,
          views: 25,
        },
        favoritedUsers: [{
          userId: 1,
        }, {
          userId: 2,
        }],
        voters: [{
          userId: 1,
          hasVoted: true,
        }],
        voteMessage: 'Upvote removed',
      },
      handleUpvote: jest.fn(),
      handleDownvote: jest.fn(),
      handleFavorite: jest.fn(),
    };
    const wrapper = shallow(<RecipeCard {...props} />);

    expect(wrapper).toBeDefined();
    expect(wrapper).toMatchSnapshot();
  });
  it('should render component with downvote', () => {
    const props = {
      singleRecipe: {
        recipe: {
          title: 'Macaroni',
          recipeImage: 'Recipe image URL',
          upvotes: 4,
          downvotes: 0,
          favorites: 8,
          views: 25,
        },
        favoritedUsers: [{
          userId: 1,
        }],
        voters: [{
          userId: 2,
          hasVoted: false,
        }, {
          userId: 3,
          hasVoted: true,
        }],
        voteMessage: 'Downvote recorded',
      },
      handleUpvote: jest.fn(),
      handleDownvote: jest.fn(),
      handleFavorite: jest.fn(),
    };
    const wrapper = shallow(<RecipeCard {...props} />);

    expect(wrapper).toBeDefined();
    wrapper.find('#downvote').simulate('click');
    expect(props.handleDownvote).toBeCalled();
    expect(wrapper).toMatchSnapshot();
  });
  it('should render component with downvote', () => {
    const props = {
      singleRecipe: {
        recipe: {
          title: 'Macaroni',
          recipeImage: 'Recipe image URL',
          upvotes: 4,
          downvotes: 0,
          favorites: 8,
          views: 25,
        },
        favoritedUsers: [{
          userId: 1,
        }],
        voters: [{
          userId: 2,
          hasVoted: false,
        }, {
          userId: 3,
          hasVoted: true,
        }],
        voteMessage: 'Downvote removed',
      },
      handleUpvote: jest.fn(),
      handleDownvote: jest.fn(),
      handleFavorite: jest.fn(),
    };
    const wrapper = shallow(<RecipeCard {...props} />);

    expect(wrapper).toBeDefined();
    expect(wrapper).toMatchSnapshot();
  });
  it('should render component with favorite', () => {
    const props = {
      singleRecipe: {
        recipe: {
          title: 'Macaroni',
          recipeImage: 'Recipe image URL',
          upvotes: 4,
          downvotes: 0,
          favorites: 8,
          views: 25,
        },
        favoritedUsers: [],
        voters: [],
        favoriteMessage: 'Recipe added to favorites',
      },
      handleUpvote: jest.fn(),
      handleDownvote: jest.fn(),
      handleFavorite: jest.fn(),
    };
    const wrapper = shallow(<RecipeCard {...props} />);

    expect(wrapper).toBeDefined();
    wrapper.find('#favorite').simulate('click');
    expect(props.handleFavorite).toBeCalled();
    expect(wrapper).toMatchSnapshot();
  });
  it('should render component with favorite', () => {
    const props = {
      singleRecipe: {
        recipe: {
          title: 'Macaroni',
          recipeImage: 'Recipe image URL',
          upvotes: 4,
          downvotes: 0,
          favorites: 8,
          views: 25,
        },
        favoritedUsers: [],
        voters: [],
        favoriteMessage: 'Recipe removed from favorites',
      },
      handleUpvote: jest.fn(),
      handleDownvote: jest.fn(),
      handleFavorite: jest.fn(),
    };
    const wrapper = shallow(<RecipeCard {...props} />);

    expect(wrapper).toBeDefined();
    wrapper.find('#favorite').simulate('click');
    expect(props.handleFavorite).toBeCalled();
    expect(wrapper).toMatchSnapshot();
  });
});
