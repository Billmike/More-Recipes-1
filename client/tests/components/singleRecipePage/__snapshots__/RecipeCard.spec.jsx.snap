// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecipeCard component should render component 1`] = `
<div
  className="card single-card hoverable"
>
  <div
    className="card-image"
  >
    <RecipeImage
      recipeImage="Recipe image URL"
      title="Macaroni"
    />
  </div>
  <div
    className="card-action center-align"
  >
    <a
      id="views"
    >
      <i
        className="material-icons black-text"
        data-tip="views"
      >
        visibility
      </i>
      <span
        className="black-text"
      >
        25
      </span>
    </a>
    <a
      className="chip waves-effect waves-red "
      data-tip="favorite"
      id="favorite"
      onClick={[MockFunction]}
      role="button"
      tabIndex="0"
    >
      <i
        className="fa fa-heart black-text"
      />
      <span
        className="black-text"
      >
        8
      </span>
    </a>
    <a
      className="chip waves-effect waves-red "
      data-tip="upvote"
      id="upvote"
      onClick={[MockFunction]}
      role="button"
      tabIndex="0"
    >
      <i
        className="fa fa-thumbs-up black-text"
      />
      <span
        className="black-text"
      >
        4
      </span>
    </a>
    <a
      className="chip waves-effect waves-red "
      data-tip="downvote"
      id="downvote"
      onClick={[MockFunction]}
      role="button"
      tabIndex="0"
    >
      <i
        className="fa fa-thumbs-down black-text"
      />
      <span
        className="black-text"
      >
        0
      </span>
    </a>
    <ReactTooltip
      insecure={true}
      resizeHide={true}
      wrapper="div"
    />
  </div>
</div>
`;

exports[`RecipeCard component should render component with downvote 1`] = `
<div
  className="card single-card hoverable"
>
  <div
    className="card-image"
  >
    <RecipeImage
      recipeImage="Recipe image URL"
      title="Macaroni"
    />
  </div>
  <div
    className="card-action center-align"
  >
    <a
      id="views"
    >
      <i
        className="material-icons black-text"
        data-tip="views"
      >
        visibility
      </i>
      <span
        className="black-text"
      >
        25
      </span>
    </a>
    <a
      className="chip waves-effect waves-red "
      data-tip="favorite"
      id="favorite"
      onClick={[MockFunction]}
      role="button"
      tabIndex="0"
    >
      <i
        className="fa fa-heart black-text"
      />
      <span
        className="black-text"
      >
        8
      </span>
    </a>
    <a
      className="chip waves-effect waves-red "
      data-tip="upvote"
      id="upvote"
      onClick={[MockFunction]}
      role="button"
      tabIndex="0"
    >
      <i
        className="fa fa-thumbs-up black-text"
      />
      <span
        className="black-text"
      >
        4
      </span>
    </a>
    <a
      className="chip waves-effect waves-red red"
      data-tip="downvote"
      id="downvote"
      onClick={
        [MockFunction] {
          "calls": Array [
            Array [],
          ],
        }
      }
      role="button"
      tabIndex="0"
    >
      <i
        className="fa fa-thumbs-down white-text"
      />
      <span
        className="white-text"
      >
        0
      </span>
    </a>
    <ReactTooltip
      insecure={true}
      resizeHide={true}
      wrapper="div"
    />
  </div>
</div>
`;

exports[`RecipeCard component should render component with downvote 2`] = `
<div
  className="card single-card hoverable"
>
  <div
    className="card-image"
  >
    <RecipeImage
      recipeImage="Recipe image URL"
      title="Macaroni"
    />
  </div>
  <div
    className="card-action center-align"
  >
    <a
      id="views"
    >
      <i
        className="material-icons black-text"
        data-tip="views"
      >
        visibility
      </i>
      <span
        className="black-text"
      >
        25
      </span>
    </a>
    <a
      className="chip waves-effect waves-red "
      data-tip="favorite"
      id="favorite"
      onClick={[MockFunction]}
      role="button"
      tabIndex="0"
    >
      <i
        className="fa fa-heart black-text"
      />
      <span
        className="black-text"
      >
        8
      </span>
    </a>
    <a
      className="chip waves-effect waves-red "
      data-tip="upvote"
      id="upvote"
      onClick={[MockFunction]}
      role="button"
      tabIndex="0"
    >
      <i
        className="fa fa-thumbs-up black-text"
      />
      <span
        className="black-text"
      >
        4
      </span>
    </a>
    <a
      className="chip waves-effect waves-red "
      data-tip="downvote"
      id="downvote"
      onClick={[MockFunction]}
      role="button"
      tabIndex="0"
    >
      <i
        className="fa fa-thumbs-down black-text"
      />
      <span
        className="black-text"
      >
        0
      </span>
    </a>
    <ReactTooltip
      insecure={true}
      resizeHide={true}
      wrapper="div"
    />
  </div>
</div>
`;

exports[`RecipeCard component should render component with favorite 1`] = `
<div
  className="card single-card hoverable"
>
  <div
    className="card-image"
  >
    <RecipeImage
      recipeImage="Recipe image URL"
      title="Macaroni"
    />
  </div>
  <div
    className="card-action center-align"
  >
    <a
      id="views"
    >
      <i
        className="material-icons black-text"
        data-tip="views"
      >
        visibility
      </i>
      <span
        className="black-text"
      >
        25
      </span>
    </a>
    <a
      className="chip waves-effect waves-red red"
      data-tip="favorite"
      id="favorite"
      onClick={
        [MockFunction] {
          "calls": Array [
            Array [],
          ],
        }
      }
      role="button"
      tabIndex="0"
    >
      <i
        className="fa fa-heart white-text"
      />
      <span
        className="white-text"
      >
        8
      </span>
    </a>
    <a
      className="chip waves-effect waves-red "
      data-tip="upvote"
      id="upvote"
      onClick={[MockFunction]}
      role="button"
      tabIndex="0"
    >
      <i
        className="fa fa-thumbs-up black-text"
      />
      <span
        className="black-text"
      >
        4
      </span>
    </a>
    <a
      className="chip waves-effect waves-red "
      data-tip="downvote"
      id="downvote"
      onClick={[MockFunction]}
      role="button"
      tabIndex="0"
    >
      <i
        className="fa fa-thumbs-down black-text"
      />
      <span
        className="black-text"
      >
        0
      </span>
    </a>
    <ReactTooltip
      insecure={true}
      resizeHide={true}
      wrapper="div"
    />
  </div>
</div>
`;

exports[`RecipeCard component should render component with favorite 2`] = `
<div
  className="card single-card hoverable"
>
  <div
    className="card-image"
  >
    <RecipeImage
      recipeImage="Recipe image URL"
      title="Macaroni"
    />
  </div>
  <div
    className="card-action center-align"
  >
    <a
      id="views"
    >
      <i
        className="material-icons black-text"
        data-tip="views"
      >
        visibility
      </i>
      <span
        className="black-text"
      >
        25
      </span>
    </a>
    <a
      className="chip waves-effect waves-red "
      data-tip="favorite"
      id="favorite"
      onClick={
        [MockFunction] {
          "calls": Array [
            Array [],
          ],
        }
      }
      role="button"
      tabIndex="0"
    >
      <i
        className="fa fa-heart black-text"
      />
      <span
        className="black-text"
      >
        8
      </span>
    </a>
    <a
      className="chip waves-effect waves-red "
      data-tip="upvote"
      id="upvote"
      onClick={[MockFunction]}
      role="button"
      tabIndex="0"
    >
      <i
        className="fa fa-thumbs-up black-text"
      />
      <span
        className="black-text"
      >
        4
      </span>
    </a>
    <a
      className="chip waves-effect waves-red "
      data-tip="downvote"
      id="downvote"
      onClick={[MockFunction]}
      role="button"
      tabIndex="0"
    >
      <i
        className="fa fa-thumbs-down black-text"
      />
      <span
        className="black-text"
      >
        0
      </span>
    </a>
    <ReactTooltip
      insecure={true}
      resizeHide={true}
      wrapper="div"
    />
  </div>
</div>
`;

exports[`RecipeCard component should render component with upvote 1`] = `
<div
  className="card single-card hoverable"
>
  <div
    className="card-image"
  >
    <RecipeImage
      recipeImage="Recipe image URL"
      title="Macaroni"
    />
  </div>
  <div
    className="card-action center-align"
  >
    <a
      id="views"
    >
      <i
        className="material-icons black-text"
        data-tip="views"
      >
        visibility
      </i>
      <span
        className="black-text"
      >
        25
      </span>
    </a>
    <a
      className="chip waves-effect waves-red "
      data-tip="favorite"
      id="favorite"
      onClick={[MockFunction]}
      role="button"
      tabIndex="0"
    >
      <i
        className="fa fa-heart black-text"
      />
      <span
        className="black-text"
      >
        8
      </span>
    </a>
    <a
      className="chip waves-effect waves-red red"
      data-tip="upvote"
      id="upvote"
      onClick={
        [MockFunction] {
          "calls": Array [
            Array [],
          ],
        }
      }
      role="button"
      tabIndex="0"
    >
      <i
        className="fa fa-thumbs-up white-text"
      />
      <span
        className="white-text"
      >
        4
      </span>
    </a>
    <a
      className="chip waves-effect waves-red "
      data-tip="downvote"
      id="downvote"
      onClick={[MockFunction]}
      role="button"
      tabIndex="0"
    >
      <i
        className="fa fa-thumbs-down black-text"
      />
      <span
        className="black-text"
      >
        0
      </span>
    </a>
    <ReactTooltip
      insecure={true}
      resizeHide={true}
      wrapper="div"
    />
  </div>
</div>
`;

exports[`RecipeCard component should render component with upvote 2`] = `
<div
  className="card single-card hoverable"
>
  <div
    className="card-image"
  >
    <RecipeImage
      recipeImage="Recipe image URL"
      title="Macaroni"
    />
  </div>
  <div
    className="card-action center-align"
  >
    <a
      id="views"
    >
      <i
        className="material-icons black-text"
        data-tip="views"
      >
        visibility
      </i>
      <span
        className="black-text"
      >
        25
      </span>
    </a>
    <a
      className="chip waves-effect waves-red "
      data-tip="favorite"
      id="favorite"
      onClick={[MockFunction]}
      role="button"
      tabIndex="0"
    >
      <i
        className="fa fa-heart black-text"
      />
      <span
        className="black-text"
      >
        8
      </span>
    </a>
    <a
      className="chip waves-effect waves-red "
      data-tip="upvote"
      id="upvote"
      onClick={[MockFunction]}
      role="button"
      tabIndex="0"
    >
      <i
        className="fa fa-thumbs-up black-text"
      />
      <span
        className="black-text"
      >
        4
      </span>
    </a>
    <a
      className="chip waves-effect waves-red "
      data-tip="downvote"
      id="downvote"
      onClick={[MockFunction]}
      role="button"
      tabIndex="0"
    >
      <i
        className="fa fa-thumbs-down black-text"
      />
      <span
        className="black-text"
      >
        0
      </span>
    </a>
    <ReactTooltip
      insecure={true}
      resizeHide={true}
      wrapper="div"
    />
  </div>
</div>
`;
